Add Web-Types for Lit framework (2.1.1)

gaetanmaisse opened the pull request:
**Description**

This PR contains a first version of the Web-Types for Lit framework: https://lit.dev/

It might not cover everything provided by Lit but provides the core features and is intended to be a 1st PR done by an external contributor. This will be a good way to check contribution steps.

**Questions to discuss**

 - How should we manage icons? Should we start with a simple icon folder inside `packages/<framework>/`?
 - I tried the publish script (in dry-run mode) locally and it failed because `@web-types/lit` doesn't exist yet on the NPM registry, is there something I can do to help on that side?
 - I'm used to having some tests/lint checks, I think JSON files format is checked before publishing, should we have this check in the PR too?

gaetanmaisse commented on packages/lit/lit@2.1.1.web-types.json:
```diff
@@ -0,0 +1,86 @@
+  "version": "2.1.1",
+  "js-types-syntax": "typescript",
+  "description-markup": "markdown",
+  "default-icon": "icons/lit.svg",
```
Related to the PR description, where should we put the icons?

piotrtomiak commented:
@gaetanmaisse Thank you for your contribution! So far I haven't been publishing anything with the icons, so I think it would make sense to create a folder `packages/lit/lit@2.1.1` and place full contents of the package inside.

@jpradelle has been doing some stuff around Web-Types for lit. Is there anything you would add to the file?

jpradelle commented on packages/lit/lit@2.1.1.web-types.json:
```diff
@@ -0,0 +1,86 @@
+          "inject-language": "JavaScript",
+          "pattern": {
+            "items": {
+              "path": "/html/attributes",
```
```suggestion
              "path": "/js/properties",
```
In lit, attributes can be named differently than properties. I think then path should be "/js/properties".

For example:
A `test-component`
```js
import {LitElement, html} from 'lit';
import {customElement, property} from 'lit/decorators.js';

@customElement('test-component')
export class TestComponent extends LitElement {
  @property({attribute: 'test-demo'})
  testDemo;

  render() {
    return html`${this.testDemo}`;
  }
}
```
Would be used with following usage:
```html
      <test-component test-demo="foo"></test-component>
      <test-component .testDemo="${this.bar}"></test-component>
```

Corresponding component web-types file would be something like
```json
{
  "$schema": "http://json.schemastore.org/web-types",
  "name": "...",
  "version": "...",
  "framework": "lit",
  "contributions": {
    "html": {
      "elements": [
        {
          "name": "test-component",
          "attributes": [
            {
              "name": "test-demo",
              "value": {
                "type": "string",
                "required": true
              }
            }
          ],
          "js": {
            "properties": [
              {
                "name": "testDemo",
                "value": {
                  "type": "string",
                  "required": true
                }
              }
            ]
          }
        }
      ]
    }
  }
}
```

jpradelle reviewed (commented):
Hi,
I made this version : https://github.com/jpradelle/lit-web-types/blob/master/web-types.json
Thanks you for this pull request, I will deprecate my package after this PR is merged.

I have a question on events: in pattern I added a generic regex to allow all values and avoid IDE warning telling event is unrecognized in case of bubbling event. This gives generic documentation "Event listener" with lit documentation link on events fired by a sub element of the dom. @piotrtomiak what would you recommend on that for bubbling events ?

My version merged with @gaetanmaisse version's:
```json
          "pattern": {
            "or": [
              {
                "items": "/js/events",
                "template": [
                  "@",
                  "$...",
                  "#item:event name"
                ]
                "priority": "normal"
              },
              {
                "template": [
                  {
                    "regex": "@[-a-z0-9]+"
                  }
                ],
                "priority": "lowest"
              }
            ]
          }
```

piotrtomiak commented:
> @piotrtomiak what would you recommend on that for bubbling events 

What do you mean by "bubbling events" - I am not an expert with frontend ;)  

@jpradelle `"/html/events"` is the same as `"/js/events"` so you don't need to include both.

jpradelle commented:
For event bubbling I won't be as good as other resources to explain it https://stackoverflow.com/questions/4616694/what-is-event-bubbling-and-capturing

I made a small demo to explain the issue [here](https://lit.dev/playground/#project=W3sibmFtZSI6ImZvby1jb21wb25lbnQudHMiLCJjb250ZW50IjoiaW1wb3J0IHtMaXRFbGVtZW50LCBodG1sfSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBwcm9wZXJ0eX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnZm9vLWNvbXBvbmVudCcpXG5leHBvcnQgY2xhc3MgRm9vQ29tcG9uZW50IGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gIEBwcm9wZXJ0eSh7dHlwZTogQm9vbGVhbn0pXG4gIGJ1YmJsZSA9IGZhbHNlO1xuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGBcbiAgICAgIDxidXR0b24gQGNsaWNrPVwiJHt0aGlzLmZpcmVFdmVudH1cIj5DbGljayBtZSAke3RoaXMuYnViYmxlID8gJyB3aXRoIGJ1YmJsZScgOiAnJ308L2J1dHRvbj5cbiAgICBgO1xuICB9XG5cbiAgZmlyZUV2ZW50KCkge1xuICAgIHRoaXMuZGlzcGF0Y2hFdmVudChuZXcgQ3VzdG9tRXZlbnQoJ2Zvby1ldmVudCcsIHtcbiAgICAgIGJ1YmJsZXM6IHRoaXMuYnViYmxlLFxuICAgICAgY29tcG9zZWQ6IHRydWVcbiAgICB9KSk7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9mb28tY29tcG9uZW50LmpzXCI-PC9zY3JpcHQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vYmFyLWNvbXBvbmVudC5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxiYXItY29tcG9uZW50PjwvYmFyLWNvbXBvbmVudD5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfSx7Im5hbWUiOiJiYXItY29tcG9uZW50LnRzIiwiY29udGVudCI6ImltcG9ydCB7TGl0RWxlbWVudCwgaHRtbH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgc3RhdGV9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcbmltcG9ydCAnLi9mb28tY29tcG9uZW50LmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ2Jhci1jb21wb25lbnQnKVxuZXhwb3J0IGNsYXNzIEJhekNvbXBvbmVudCBleHRlbmRzIExpdEVsZW1lbnQge1xuICBAc3RhdGUoKVxuICBsb2dzID0gW107XG5cbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYFxuICAgICAgPGRpdiBAZm9vLWV2ZW50PVwiJHt0aGlzLmZvb0V2ZW50QnViYmxlfVwiPlxuICAgICAgICA8Zm9vLWNvbXBvbmVudCBAZm9vLWV2ZW50PVwiJHt0aGlzLmZvb0V2ZW50fVwiPjwvZm9vLWNvbXBvbmVudD5cbiAgICAgICAgPGZvby1jb21wb25lbnQgYnViYmxlIEBmb28tZXZlbnQ9XCIke3RoaXMuZm9vRXZlbnR9XCI-PC9mb28tY29tcG9uZW50PlxuICAgICAgPC9kaXY-XG4gICAgICAgIFxuICAgICAgPGRpdj5cbiAgICAgICAgTG9ncyA6PGJyLz5cbiAgICAgICAgJHt0aGlzLmxvZ3MubWFwKGwgPT4gaHRtbGAke2x9PGJyLz5gKX1cbiAgICAgIDwvZGl2PlxuICAgIGA7XG4gIH1cblxuICBmb29FdmVudCgpIHtcbiAgICB0aGlzLmxvZ3MgPSBbLi4udGhpcy5sb2dzLCAnUmVjaWV2ZWQgZm9vLWV2ZW50J107XG4gIH1cblxuICBmb29FdmVudEJ1YmJsZSgpIHtcbiAgICB0aGlzLmxvZ3MgPSBbLi4udGhpcy5sb2dzLCAnUmVjaWV2ZWQgZm9vLWV2ZW50IGFzIGV2ZW50IGJ1YmJsZXMnXTtcbiAgfVxufVxuIn1d)
In the demo when you click the button it fires an event, one bubbling and the other not, you will see difference in displayed logs.
The point which impact us for web-types is around that part:
```html
      <div @foo-event="${this.fooEventBubble}">
        <foo-component @foo-event="${this.fooEvent}"></foo-component>
        <foo-component bubble @foo-event="${this.fooEvent}"></foo-component>
      </div>
```
If `foo-component` fires `foo-event` with bubbling upper `div` can listen on `foo-event`. (bubbling will continue until it reach root node of the dom). If `foo-event` doesn't bubble, `div` won't receive it.

Currently on IntelliJ, with this PR web-types definition for lit, and following webcomponent web-types definition, that's gives this result:
![image](https://user-images.githubusercontent.com/24655340/152160360-68be3333-57a6-4d1c-9269-3d88b581b7c9.png)
`foo-event` is not recognized on div, but it will exist.

When using following following web-types definition
```json
{
    "$schema": "http://json.schemastore.org/web-types",
    "name": "web-types-test",
    "version": "0.1.0",
    "framework": "lit",
    "contributions": {
        "html": {
            "elements": [
                {
                    "name": "bar-component",
                    "js": {
                        "properties": [
                            {
                                "name": "logs",
                                "required": false,
                                "value": {
                                    "type": "array",
                                    "required": true,
                                    "default": "[]"
                                }
                            }
                        ]
                    }
                },
                {
                    "name": "foo-component",
                    "attributes": [
                        {
                            "name": "bubble",
                            "required": false,
                            "value": {
                                "type": "boolean",
                                "required": false,
                                "default": "false"
                            }
                        }
                    ],
                    "js": {
                        "properties": [
                            {
                                "name": "bubble",
                                "required": false,
                                "value": {
                                    "type": "boolean",
                                    "required": true,
                                    "default": "false"
                                }
                            }
                        ],
                        "events": [
                            {
                                "name": "foo-event"
                            }
                        ]
                    }
                }
            ]
        }
    }
}
```

When listening an event with bubbling, the event can come from a very far away place from which it is listened. [Composed](https://developer.mozilla.org/en-US/docs/Web/API/Event/composed) concept also enter in the game for a greater complexity. Which can make it very complicated on your side.

Alternatively, I could build web-types file differently to define event globally and have it understood by IDE, but this tells event exists everywhere, where it not really the case:
```json
{
    "$schema": "http://json.schemastore.org/web-types",
    "name": "web-types-test",
    "version": "0.1.0",
    "framework": "lit",
    "contributions": {
        "html": {
            "elements": [
                {
                    "name": "bar-component",
                    "js": {
                        "properties": [
                            {
                                "name": "logs",
                                "required": false,
                                "value": {
                                    "type": "array",
                                    "required": true,
                                    "default": "[]"
                                }
                            }
                        ]
                    }
                },
                {
                    "name": "foo-component",
                    "attributes": [
                        {
                            "name": "bubble",
                            "required": false,
                            "value": {
                                "type": "boolean",
                                "required": false,
                                "default": "false"
                            }
                        }
                    ],
                    "js": {
                        "properties": [
                            {
                                "name": "bubble",
                                "required": false,
                                "value": {
                                    "type": "boolean",
                                    "required": true,
                                    "default": "false"
                                }
                            }
                        ]
                    }
                }
            ]
        },
        "js": {
            "events": [
                {
                    "name": "foo-event"
                }
            ]
        }
    }
}
```

piotrtomiak commented:
@jpradelle To correctly handle event bubbling I think one would need to write a plugin, which based on the file contents would add additional events on the elements. I think the current solution is good enough. 

@gaetanmaisse It would be great to finish this PR and deliver these web-types automatically to all IDE users.

gaetanmaisse commented:
@piotrtomiak I was quite busy these days but I will take some time to review the comments done and update the PR 😉

gaetanmaisse commented:
@jpradelle thanks for the detailed feedback! I updated the Web-Types file according to them, can you take it a look and lmk if you think it could be a good first version?

@piotrtomiak as you proposed in a previous comment to handle some extra files I used folder structure, for instance:
<img width="232" alt="image" src="https://user-images.githubusercontent.com/4112568/156443522-c8cdf1d0-bc4a-431e-8db1-58d6b302e115.png">
To be able to publish this kind of package I updated the publication script, it _should_ handle both file and folder structure. If it's ok for you I think we will be able to publish the first community-written Web-Types file 😁

piotrtomiak commented:
@gaetanmaisse Thank you for the changes! I will update the web-types file according to @jpradelle remarks.

piotrtomiak commented:
I have also published the package. It looks like IDE is not automatically downloading it, so you must explicitly install it ("@web-types/lit"), similarly to "@types" packages.
